### PR TITLE
Move files external to core

### PR DESCRIPTION
**Pull request by PVince81**
For https://github.com/owncloud/core/issues/18160

There's more to come.

### CODE REVIEW

**Comment by PVince81** [comment]
Fixed some tests.
Moved BackendService to core as `\OCP\Files\External\IStoragesBackendService` implemented as `\OC\Files\External\StoragesBackendService`.

Looking good so far but there is more to go to have core being able to work without files_external enabled.

**Comment by PVince81** [comment]
- [ ] get rid of StorageMigrator

**Comment by PVince81** [comment]
- [ ] get rid of LegacyStoragesService: at this stage everybody already migrated their config

**Comment by PVince81** [comment]
- [ ] need alias for backend identifiers because they contain "Files_External" in them

**Comment by PVince81** [comment]
- [ ] retest all commands, including importing mount.json (legacy config)

**Comment by PVince81** [comment]
Moved storage services classes to core, still WIP/incomplete. Not tested.
Stay tuned...

**Comment by DeepDiver1975** [comment]
Conflicting

**Comment by PVince81** [comment]
Rebased, added more fixes for the tests etc. Seems to work so far, few tests will fail.

I think we should also move the commands and the UI to core.
The UI would go under "settings/".

**Comment by PVince81** [comment]
I'd say the first main goal is to make external apps like "files_external_ftp" work even when files_external is disabled. Once this works, it's already a big step forward even if a lot of things still need to be moved like the UI and CLI commands.

**Comment by PVince81** [comment]
Some people disable files _external even though they have storages configured, to disable them.
When migrating, we don't want those to suddenly pop up again.
So I suggest to first implement the feature "disable single storage" to avoid a loss of functionality and update surprises.
Then the migration code will detect whether files_external was disabled and if yes will disable all storages.

- [ ] disable all ext storages if files_external was disabled during update

**Comment by PVince81** [comment]
- [ ] find out what `setLegacyAuthMechanism*` methods are for and see if we can get rid of them

**Comment by PVince81** [comment]
Good news! It is now possible to run this FTP branch https://github.com/owncloud/files_external_ftp/pull/9 with this PR here and have the "files_external" app disabled !

You'll still need to enable "files_external" first to configure the storage but then can disable it again. The storage will stay available.

There is still a lot of work to do to make sure everything still works.

And also review whether each class I picked for the public or private namespace makes sense.
I had to add a few aliases in the server container for DI to work properly.

**Comment by PVince81** [comment]
```
1) Test\Files\External\Service\GlobalStoragesServiceTest::testUpdateStorage with data set #0 (array('mountpoint', 'identifier:\Test\Files\Extern...ackend', 'identifier:\Auth\Mechanism', array('value1', 'value2', 'testPassword'), array(), array(), 15))
18:15:43 TypeError: Argument 1 passed to Mock_AuthMechanism_86ad8386::wrapStorage() must implement interface OCP\Files\Storage, null given, called in /var/lib/jenkins/workspace/owncloud-core_core_PR-25422-NVXOMA4CC3FS3H34SQTLTEWMW4QUKUB4WWGYUPAFM4P2XYM3IARA/lib/private/Files/External/Service/StoragesService.php on line 535
18:15:43 
18:15:43 /var/lib/jenkins/workspace/owncloud-core_core_PR-25422-NVXOMA4CC3FS3H34SQTLTEWMW4QUKUB4WWGYUPAFM4P2XYM3IARA/lib/private/Files/External/Service/StoragesService.php:535
18:15:43 /var/lib/jenkins/workspace/owncloud-core_core_PR-25422-NVXOMA4CC3FS3H34SQTLTEWMW4QUKUB4WWGYUPAFM4P2XYM3IARA/lib/private/Files/External/Service/StoragesService.php:451
18:15:43 /var/lib/jenkins/workspace/owncloud-core_core_PR-25422-NVXOMA4CC3FS3H34SQTLTEWMW4QUKUB4WWGYUPAFM4P2XYM3IARA/tests/lib/Files/External/Service/GlobalStoragesServiceTest.php:180
18:15:43 
18:15:43 2) Test\Files\External\Service\GlobalStoragesServiceTest::testUpdateStorage with data set #1 (array('mountpoint', 'identifier:\Test\Files\Extern...ackend', 'identifier:\Auth\Mechanism', array('value1', 'value2', 'testPassword'), array('user1', 'user2'), array(), 15))
18:15:43 TypeError: Argument 1 passed to Mock_AuthMechanism_86ad8386::wrapStorage() must implement interface OCP\Files\Storage, null given, called in /var/lib/jenkins/workspace/owncloud-core_core_PR-25422-NVXOMA4CC3FS3H34SQTLTEWMW4QUKUB4WWGYUPAFM4P2XYM3IARA/lib/private/Files/External/Service/StoragesService.php on line 535
18:15:43 
```

Weird, I don't get these locally... Maybe related to PHP version.

**Comment by PVince81** [comment]
Aha! Indeed I see these errors with PHP 7.0.11

**Comment by PVince81** [comment]
- [ ] get rid of `OC_MountConfig` references in core:
```
lib/private/Files/External/Service/LegacyStoragesService.php:82:                if ($mountType === \OC_Mount_Config::MOUNT_TYPE_USER) {
lib/private/Files/External/Service/LegacyStoragesService.php:88:                } else if ($mountType === \OC_Mount_Config::MOUNT_TYPE_GROUP) {
lib/private/Files/External/Service/LegacyStoragesService.php:153:                                       $storageOptions['options'] = \OC_Mount_Config::decryptPasswords($storageOptions['opti
...
```

**Comment by PVince81** [comment]
- [ ] oh, and fix encryption:
```
apps/encryption/lib/Migration.php:323:          return \OC_Mount_Config::getSystemMountPoints();
lib/private/Encryption/File.php:88:                     $mounts = \OC_Mount_Config::getSystemMountPoints();
lib/private/Encryption/Util.php:290:                    $mounts = \OC_Mount_Config::getSystemMountPoints();
```

**Comment by PVince81** [comment]
- [ ] raise ticket for future: ability to register mount providers / storage backends in info.xml

**Comment by PVince81** [comment]
@DeepDiver1975 @jvillafanez should we get rid of the ability to import old style "mount.json" using the "Import" command ? This would make it possible to clear a lot of legacy cruft and reduce testing overhead.

The only use case I can think of is admins who provision their OC using mount.json files automatically and haven't taken the time to convert these configs to the new json export format.

As for existing mount.json files inside the install, these should already be migrated to the DB format.

**Comment by PVince81** [comment]
Needs retesting:
- [ ] TEST: existing storage apps that still rely on OCA\Files_external namespace, make sure they still work
- [ ] TEST: system mounts
- [ ] TEST: personal mounts
- [ ] TEST: admin page settings
- [ ] TEST: personal page settings
- [ ] TEST: status manager no errors
- [ ] TEST: all CLI commands
- [ ] TEST: encryption with system mounts applicable to groups + users (it uses external storage API to find which users have access)
- [ ] TEST: encryption with personal mounts
- [ ] TEST: ftp app with files_external disabled

**Comment by PVince81** [comment]
- [ ] add shims for password/null/builtin auth mechanisms in case old apps are using it ?
- [ ] move Password auth mechanism to public namespace to allow storages using it as legacy auth ?

**Comment by PVince81** [comment]
- [ ] BUG: order of auth mechanisms is now different. "Username and password" should be at the top in the dropdown. (this is due to registration order...)

**Comment by PVince81** [comment]
- [ ] provide interface of the `StorageServices` classes, seems there is a need for these in WND for example

**Comment by PVince81** [comment]
- [ ] quick doc about porting old storages:
    - search for all uses for "OCA\Files_External" and replace to either "OCP\Files\External" (public) or "OC\Files\External" (private, should not use)
    - StorageConfig -> IStorageConfig
    - get rid of `OC_Mount_Config` and use constants from other interfaces. In worst case use "\OC\Files\External\LegacyUtil"

**Comment by PVince81** [comment]
some performance concerns: after merging this every setupFS call will have to read the ext storage tables even if there is nothing there. So even people who don't use ext storage would get a perf decrease. Possible long-term solution here in the fstab discussion: https://github.com/owncloud/core/issues/26190

**Comment by PVince81** [comment]
Failure possibly due to array expectation order:
```
13:32:03 1) Test\Files\External\Service\DBConfigServiceTest::testAddApplicable
13:32:03 Failed asserting that two arrays are equal.
13:32:03 --- Expected
13:32:03 +++ Actual
13:32:03 @@ @@
13:32:03  Array (
13:32:03      0 => Array (
13:32:03 -        'type' => 3
13:32:03 -        'value' => 'test'
13:32:03 -        'mount_id' => 2
13:32:03 +        'type' => 1
13:32:03 +        'value' => null
13:32:03 +        'mount_id' => '2'
13:32:03      )
13:32:03      1 => Array (...)
13:32:03      2 => Array (
13:32:03 -        'type' => 1
13:32:03 -        'value' => null
13:32:03 -        'mount_id' => 2
13:32:03 +        'type' => 3
13:32:03 +        'value' => 'test'
13:32:03 +        'mount_id' => '2'
13:32:03      )
13:32:03  )
13:32:03 
13:32:03 /var/lib/jenkins/workspace/owncloud-core_core_PR-25422-NVXOMA4CC3FS3H34SQTLTEWMW4QUKUB4WWGYUPAFM4P2XYM3IARA/tests/lib/Files/External/Service/DBConfigServiceTest.php:96
13:32:03 
```

**Comment by butonic** [comment]
will fix https://github.com/owncloud/core/issues/22641

**Comment by PVince81** [comment]
Raised https://github.com/owncloud/core/issues/26538 to handle missing backends more gracefully.

**Comment by PVince81** [comment]
This is tricky... I tried enabling the WND app and making it work without any modifications to the app.
This is not possible with this PR because we need to introduce proper interfaces and also the WND app is using classes from OCA\Files_External that it's not supposed to use.

And the WND app also provides some additional auth providers which need to be adjusted to use `IStorageConfig` instead of the old object to be compatible. I guess it's ok to ask storage and auth providers to adjust their implementations to this PR's change.

**Comment by DeepDiver1975** [comment]
> And the WND app also provides some additional auth providers which need to be adjusted to use IStorageConfig instead of the old object to be compatible. I guess it's ok to ask storage and auth providers to adjust their implementations to this PR's change.

absolutly

**Comment by PVince81** [comment]
Tested updating from OC 8.0 with SFTP_Key, SMB_OC, Dropbox and others up to this branch, all worked fine. This means that getting rid of the legacy stuff did not break those, which is great news.

**Comment by PVince81** [comment]
What sound does it make when squashing 21 commits ?

Not sure, but that's what I did here. Hopefully a single commit will make all the renames more apparent.

**Comment by PVince81** [comment]
Went through my test steps and it's looking good so far !

Apart from this little bug https://github.com/owncloud/core/pull/25422#issuecomment-251910416 I think this could be merged.

However once this is merge we should work on making it fail gracefully in case a third party ext storage app gets disabled during update, this would be handled through an approach like this: https://github.com/owncloud/core/pull/26539

**Comment by jvillafanez** [comment]
We should make the Backend class abstract with an abstract method "initializeBackend". Something like:
```
abstract class Backend implements \JsonSerializable {
    public function __construct(IL10N $a) {
        $this->initializeBackend($a);
    }
    abstract protected function initializeBackend(IL10N $a);
}
```
This way we can force the Backend class to not be instanciated (it doesn't make any sense), and also point out to the subclasses what is the method that they should implement.

If the subclasses choose to redefine the constructor, they should call the parent constructor first (which is common practice anyway)

File: apps/files_external/lib/Lib/Backend/SFTP_Key.php [hunk 1]
```diff
@@ -42,7 +40,6 @@ public function __construct(IL10N $l, RSA $legacyAuth, SFTP $sftpBackend) {
 					->setFlag(DefinitionParameter::FLAG_OPTIONAL),
 			])
 			->addAuthScheme(AuthMechanism::SCHEME_PUBLICKEY)
-			->setLegacyAuthMechanism($legacyAuth)
 			->deprecateTo($sftpBackend)
```
Comment (jvillafanez) [inline apps/files_external/lib/Lib/Backend/SFTP_Key.php]: I'm not finding this method... 😕

Comment (PVince81) [inline apps/files_external/lib/Lib/Backend/SFTP_Key.php]: It's in `IdentifierTrait`

**Comment by PVince81** [comment]
> This way we can force the Backend class to not be instanciated (it doesn't make any sense), and also point out to the subclasses what is the method that they should implement.

While I agree with the base idea there are some issues:

1) backends might need dependency injection in their constructors. So I'd say the init method should have no arguments. It's the backend's responsibility to get the auto-DI from the constructor and store the instances. => no big issue

2) Instantiation is needed to get access to the param definition. Not sure if always needed in a regular usage where the admin page is not shown, would need to check.

3) When instantiating `StorageConfig` we are storing an instance of `Backend` quite early. Ideally I'd like to change this to only store the backend id and only instantiate the backend (and also auth mechanisms) at the last moment. But I'd say out of scope for this PR.

@jvillafanez what do you think ? Is what I mentionned in 3) also what you hoped by making the init lazy ?

**Comment by PVince81** [comment]
Ref from the old code: https://github.com/owncloud/core/blob/v9.1.2/apps/files_external/lib/Service/StoragesService.php#L295

**Comment by jvillafanez** [comment]
For me it doesn't make any sense that we can create instances of the Backend class when we have specific implementations. That's the reason to make the Backend class abstract.

I didn't think about lazy initialization... I completely agree we should go that way. The only we should take care is that we'll need to make this completely transparent to the implementations: if we consider that the subclasses must implement an "initBackend" method for example, checking if the backend is initialized or "deinitialize" the backend should be done transparently in the parent class.

**Comment by PVince81** [comment]
@jvillafanez ok, now I understand. I'm fine making the Backend class abstract. Not sure if `initBackend` is needed here because subclasses can still have their own constructor and we're not trying to enforce any specific signature.

**Comment by jvillafanez** [comment]
I don't think it's worthy to initialize lazily the backend. Too much trouble for a small benefit in this case since most of the things the backend does are assigments. Let's just make the class abstract.

**Comment by PVince81** [comment]
Here, they're abstract now: https://github.com/owncloud/core/pull/25422/commits/e33028ce0cb35b26ec586995a7b49f23d1d8e171

Let's hope the tests still pass, if they do, merge ? @jvillafanez @DeepDiver1975

**Comment by DeepDiver1975** [comment]
:+1:

**Pull request merged**